File: app_manager/appmanager/src/main/java/app/editors/manager/ui/dialogs/AccountBottomDialog.java
package app.editors.manager.ui.dialogs;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import moxy.presenter.InjectPresenter;
import com.google.android.material.bottomsheet.BottomSheetDialogFragment;

import java.util.List;

import app.editors.manager.R;
import app.editors.manager.app.WebDavApi;
import app.editors.manager.mvp.models.account.AccountsSqlData;
import app.editors.manager.mvp.presenters.login.AccountsPresenter;
import app.editors.manager.mvp.views.login.AccountsView;
import app.editors.manager.ui.activities.login.PortalsActivity;
import app.editors.manager.ui.activities.login.SignInActivity;
import app.editors.manager.ui.activities.login.WebDavLoginActivity;
import app.editors.manager.ui.activities.main.MainActivity;
import app.editors.manager.ui.adapters.BottomAccountAdapter;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;
import lib.toolkit.base.ui.dialogs.base.BaseBottomDialog;

public class AccountBottomDialog extends BaseBottomDialog implements BottomAccountAdapter.OnItemClickListener,
        AccountsView {

    public static final String TAG = AccountBottomDialog.class.getSimpleName();

    public static AccountBottomDialog newInstance() {
        return new AccountBottomDialog();
    }

    @BindView(R.id.accountsRecyclerView)
    RecyclerView mRecyclerView;

    @InjectPresenter
    AccountsPresenter mAccountsPresenter;

    private Unbinder mUnbinder;
    private BottomAccountAdapter mAdapter;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setStyle(BottomSheetDialogFragment.STYLE_NO_FRAME, R.style.ContextMenuDialog);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        mAdapter.setOnAddAccountClick(null);
        if (mUnbinder != null) {
            mUnbinder.unbind();
        }
    }

    @Nullable
    @Override
    @SuppressLint("InflateParams")
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.bottom_fragment_accounts, null);
        mUnbinder = ButterKnife.bind(this, view);
        return view;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        init();
    }

    private void init() {
        initPortalsAccounts();
    }

    private void initPortalsAccounts() {
        mAdapter = new BottomAccountAdapter(requireContext());
        mRecyclerView.setLayoutManager(new LinearLayoutManager(requireContext()));
        mRecyclerView.setAdapter(mAdapter);
        mRecyclerView.setNestedScrollingEnabled(true);
        mAdapter.setOnItemClickListener(this);
        mAdapter.setOnAddAccountClick(() -> PortalsActivity.showPortals(getActivity()));
        mAccountsPresenter.getAccounts();
    }

    @Override
    public void onItemClick(View view, int position) {
        mAccountsPresenter.setAccountClicked(mAdapter.getItem(position), position);
        mAccountsPresenter.loginAccount();
    }

    @Override
    public void onAccountLogin() {
        hideDialog();
        if (getContext() != null && getActivity() != null) {
            final Intent intent = new Intent(getContext(), MainActivity.class);
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N && getActivity().isInMultiWindowMode()) {
                intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK | Intent.FLAG_ACTIVITY_LAUNCH_ADJACENT);
            }
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
            getContext().startActivity(intent);
        }
    }

    @Override
    public void onUsersAccounts(List<AccountsSqlData> accounts) {
        mAdapter.setItems(accounts);
    }

    @Override
    public void onAccountDelete(int position) {

    }

    @Override
    public void onSignIn(String portal, String login) {
        hideDialog();
        SignInActivity.showPortalSignIn(getContext(), portal, login);
    }

    @Override
    public void showWaitingDialog() {
        showWaitingDialog(getString(R.string.dialogs_wait_title),
                getString(R.string.dialogs_common_cancel_button), TAG);;
    }

    @Override
    public void onWebDavLogin(AccountsSqlData account) {
        WebDavLoginActivity.show(getActivity(), WebDavApi.Providers.valueOf(account.getWebDavProvider()), account);
    }

    @Override
    public void onError(@Nullable String message) {
        hideDialog();
        if (message != null) {
            Toast.makeText(requireContext(), message, Toast.LENGTH_LONG).show();
        }
    }
}

File: app_manager/appmanager/src/main/java/app/editors/manager/ui/dialogs/AccountBottomDialog.kt
package app.editors.manager.ui.dialogs

import android.annotation.SuppressLint
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import app.documents.core.account.CloudAccount
import app.documents.core.network.ApiContract
import app.documents.core.webdav.WebDavApi
import app.editors.manager.R
import app.editors.manager.dropbox.ui.fragments.DropboxSignInFragment
import app.editors.manager.googledrive.ui.fragments.GoogleDriveSignInFragment
import app.editors.manager.managers.utils.Constants
import app.editors.manager.mvp.models.account.Storage
import app.editors.manager.mvp.presenters.login.AccountsPresenter
import app.editors.manager.mvp.views.login.AccountsView
import app.editors.manager.onedrive.managers.utils.OneDriveUtils
import app.editors.manager.onedrive.ui.fragments.OneDriveSignInFragment
import app.editors.manager.ui.activities.login.PortalsActivity
import app.editors.manager.ui.activities.login.SignInActivity
import app.editors.manager.ui.activities.login.WebDavLoginActivity
import app.editors.manager.ui.activities.main.MainActivity
import app.editors.manager.ui.adapters.BottomAccountAdapter
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import lib.toolkit.base.managers.utils.FragmentUtils.showFragment
import lib.toolkit.base.ui.adapters.BaseAdapter
import lib.toolkit.base.ui.dialogs.base.BaseBottomDialog
import moxy.presenter.InjectPresenter

class AccountBottomDialog : BaseBottomDialog(), BaseAdapter.OnItemClickListener, AccountsView {

    companion object {
        val TAG: String = AccountBottomDialog::class.java.simpleName

        fun newInstance(): AccountBottomDialog {
            return AccountBottomDialog()
        }
    }

    @InjectPresenter
    lateinit var presenter: AccountsPresenter

    private lateinit var recyclerView: RecyclerView
    private lateinit var adapter: BottomAccountAdapter

    private var toast: Toast? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setStyle(STYLE_NO_FRAME, lib.toolkit.base.R.style.ContextMenuDialog)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        adapter.setOnAddAccountClick(null)
    }

    override fun onPause() {
        super.onPause()
        toast?.cancel()
    }

    @SuppressLint("InflateParams")
    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        return inflater.inflate(R.layout.bottom_fragment_accounts, null).apply {
            recyclerView = findViewById(R.id.accountsRecyclerView)
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        init()
    }

    private fun init() {
        initPortalsAccounts()
    }

    private fun initPortalsAccounts() {
        adapter = BottomAccountAdapter()
        recyclerView.layoutManager = LinearLayoutManager(requireContext())
        recyclerView.adapter = adapter
        recyclerView.isNestedScrollingEnabled = true
        adapter.setOnItemClickListener(this)
        adapter.setOnAddAccountClick { PortalsActivity.showPortals(requireActivity()) }
        presenter.accounts
    }

    override fun onItemClick(view: View, position: Int) {
        presenter.setAccountClicked(adapter.getItem(position), position)
        presenter.loginAccount()
    }

    override fun onAccountLogin() {
        hideDialog()
        if (context != null && activity != null) {
            val intent = Intent(context, MainActivity::class.java)
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N && requireActivity().isInMultiWindowMode) {
                intent.flags =
                    Intent.FLAG_ACTIVITY_CLEAR_TASK or Intent.FLAG_ACTIVITY_LAUNCH_ADJACENT
            }
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK)
            requireActivity().startActivity(intent)
        }
    }

    override fun onUsersAccounts(accounts: MutableList<CloudAccount>) {
        adapter.setItems(accounts)
    }

    override fun onAccountDelete(position: Int) {

    }

    override fun onSignIn(portal: String, login: String) {
        hideDialog()
        SignInActivity.showPortalSignIn(this, portal, login)
    }

    override fun showWaitingDialog() {
        showWaitingDialog(
            getString(R.string.dialogs_wait_title),
            getString(R.string.dialogs_common_cancel_button), TAG
        )
    }

    override fun onWebDavLogin(account: CloudAccount) {
        WebDavLoginActivity.show(requireActivity(), WebDavApi.Providers.valueOf(account.webDavProvider ?: ""), Json.encodeToString(account))
    }

    override fun onOneDriveLogin() {
        val storage = Storage(
            OneDriveUtils.ONEDRIVE_STORAGE,
            Constants.OneDrive.COM_CLIENT_ID,
            Constants.OneDrive.COM_REDIRECT_URL
        )
        fragmentManager?.let { showFragment(fragmentManager = it,OneDriveSignInFragment.newInstance(storage), R.id.frame_container, OneDriveSignInFragment.TAG, false) }
    }

    override fun onDropboxLogin() {
        val storage = Storage(
            ApiContract.Storage.DROPBOX,
            Constants.DropBox.COM_CLIENT_ID,
            Constants.DropBox.COM_REDIRECT_URL
        )
        fragmentManager?.let { showFragment(fragmentManager = it,DropboxSignInFragment.newInstance(storage), R.id.frame_container, DropboxSignInFragment.TAG, false) }
    }

    override fun onGoogleDriveLogin() {
        val storage = Storage(
            ApiContract.Storage.GOOGLEDRIVE,
            Constants.Google.COM_CLIENT_ID,
            Constants.Google.COM_REDIRECT_URL
        )
        fragmentManager?.let { showFragment(fragmentManager = it, GoogleDriveSignInFragment.newInstance(storage), R.id.frame_container, GoogleDriveSignInFragment.TAG, false) }
    }

    override fun onError(message: String?) {
        hideDialog()
        if (message != null) {
            toast?.cancel()
            toast = Toast.makeText(requireContext(), message, Toast.LENGTH_SHORT)
            toast?.show()
        }
    }
}